IMAP-Science-Operations-Center/sds-data-manager
# Processing Trigger Mechanisms

# Change Summary

## Overview
The most important part of this PR is what happens in batch_starter.py. It does the following:
1. Get details of the object dropped in the s3 bucket (level, version, process_dates). 
2. Query dependents of the instrument (who had an object dropped in its bucket) to see if they have been ingested
3. Decide if we have sufficient dependencies for each dependent to process. 
4. If we have dependencies to process, then send relevant information to the Step Function. 

This is for normal processing. I have also thought about how we would reprocess:
1. Manually submit relevant information to Step Function.
2. If we only want to process a single Step Function we would manually turn off the Rules for instrument dependencies for EventBridge.

## New Files
**instruments/batch_starter.py** : lambda that queries the database to decide whether or not a step function should kick off and what information to send it.
**instruments/dependents.py** : My best shot at creating a json of dependents based on diagram Tenzin created.
**instruments/Dockerfile** : Creates environment for Dockerized lambda.

**utils/dependencies.json** : My best shot at creating a json of dependencies based on diagram Tenzin created.
**utils/get_dependency.py** : Gets instrument dependents based on dependencies.json.

## Deleted Files
**lambda_code/SDSCode/instruments/l1b_codice.py** : we only need a single lambda for checking the database.
**lambda_code/SDSCode/instruments/l1c_codice.py**: we only need a single lambda for checking the database.

## Updated Files
**batch_compute_resources** : remove unnecessary batch_security_group
**instrument_lambdas**: dockerize lambda; improve lambda environment
**sdc_step_function**: remove lambda, change what is passed to batch job
**database_stack**: minor changes to allow rds_creds to be passed to other stacks. This allowed batch_starter lambda to access the database.
**networking_stack**: added endpoint so that lambda could access secrets manager
**processing_stack**: updated the eventbridge rules
**stackbuilder**: reformatting

## Testing
**lambda_endpoints/test_batch_starter.py** : Creates mock database and mock event that can be used for testing functions in batch_starter.py
**test-data/codicehi_event.json** : Contains example of json created by an event
**process_kickoff_test_data.csv** : Contains example of how an instrument table in the database may be setup

To test after deployment:
1. Create directory in s3 bucket: sds-data-<acct #>/CodiceHi/l1a/ and place an object in that bucket
2. Navigate to CloudWatch > log groups > InstrumentLambdas and click on logstream. This is the error you should see. That shows that the lambda connected to the database which is not yet populated with schema (sdc) and table (codicehi).

[ERROR] UndefinedTable: relation "sdc.codicehi" does not exist

## Notes
1. I decided to use a single lambda for all instruments since they would all have common tools. The lambda does not include any lookup for calibration tables, but I thought that might be performed in the batch job instead.
2. Likely needs many more tests once we have RDS setup and code for Docker image for Batch Job complete.

Dependency diagram used:

![Common Processing steps (Draft copies)](https://github.com/IMAP-Science-Operations-Center/sds-data-manager/assets/46567335/32fde3a6-41c1-48be-8e88-981e4cf7307c)
